File: kotlin/src/main/kotlin/advertisement/domain/model/Advertisement.kt
package advertisement.domain.model

import advertisement.domain.model.value_object.AdvertisementDate
import advertisement.domain.model.value_object.AdvertisementId
import advertisement.domain.model.value_object.Description
import advertisement.domain.model.value_object.Password
import arrow.core.Either
import arrow.core.flatMap
import common.BoundedContextError
import java.time.LocalDateTime

class Advertisement private constructor(val id: AdvertisementId, var description: Description, var password: Password, var date: AdvertisementDate)
{
    // We use flatmap instead of Bind to show the difference

    companion object
    {
        fun build(id: String, description: String, password: Password, date: LocalDateTime): Either<BoundedContextError, Advertisement>
        {
            return AdvertisementId.build(id).flatMap {
                advertisementId -> Description.build(description).flatMap {
                    description -> AdvertisementDate.build(date).map {
                        advertisementDate -> Advertisement(advertisementId, description, password, advertisementDate)
                    }
                }
            }
        }
    }

    fun update(description: Description, password: Password): Either<BoundedContextError, Advertisement> {
        this.description = description
        this.password = password

        return updateDate().map {
            this
        }
    }

    fun renew(password: Password): Either<BoundedContextError, Advertisement>
    {
        this.password = password

        return updateDate().map {
            this
        }
    }

    private fun updateDate():  Either<BoundedContextError, AdvertisementDate>
    {
        return AdvertisementDate.build(LocalDateTime.now()).map {
            this.date = it
            it
        }
    }
}

File: typescript/src/advertisement/domain/model/Advertisement.ts
import {Password} from "./value-object/Password";
import {Description} from "./value-object/Description";
import {AdvertisementId} from "./value-object/AdvertisementId";
import {AdvertisementDate} from "./value-object/AdvertisementDate";
import {DomainException} from "../../../common/domain/DomainException";
import * as E from '@effect-ts/core/Either';

export class Advertisement {

  private constructor(
    private readonly _id: AdvertisementId,
    private _description: Description,
    private _password: Password,
    private _date: AdvertisementDate
  ) {
  }

  public static build(id: string, description: string, password: Password, date: Date): E.Either<DomainException, Advertisement> {
    const advertisementIdResult = AdvertisementId.build(id);
    if (advertisementIdResult._tag === 'Left') {
      return E.left(advertisementIdResult.left);
    }
    const descriptionResult = Description.build(description);
    if (descriptionResult._tag === 'Left') {
      return E.left(descriptionResult.left);
    }
    const advertisementDateResult = AdvertisementDate.build(date);
    if (advertisementDateResult._tag === 'Left') {
      return E.left(advertisementDateResult.left);
    }
    return E.right(new Advertisement(
      advertisementIdResult.right,
      descriptionResult.right,
      password,
      advertisementDateResult.right
    ));
  }

  public update(description: Description, password: Password): E.Either<DomainException, Advertisement> {
    this._description = description;
    this._password = password;

    const result = this.updateDate();
    if (result._tag === 'Left') {
      return E.left(result.left);
    }

    return E.right(this);
  }

  public renew(password: Password): E.Either<DomainException, Advertisement> {
    this._password = password;

    const result = this.updateDate();
    if (result._tag === 'Left') {
      return E.left(result.left);
    }

    return E.right(this);
  }

  public id(): AdvertisementId {
    return this._id
  }

  public description(): Description {
    return this._description
  }

  public password(): Password {
    return this._password
  }

  public date(): AdvertisementDate {
    return this._date
  }

  private updateDate(): E.Either<DomainException, Advertisement> {
    const advertisementDateResult = AdvertisementDate.build(new Date());
    if (advertisementDateResult._tag === 'Left') {
      return E.left(advertisementDateResult.left);
    }

    this._date = advertisementDateResult.right;

    return E.right(this);
  }
}
